include/maths/CBoostedTreeFactory.h

```diff
@@ -0,0 +1,113 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+#ifndef INCLUDED_ml_maths_CBoostedTreeFactory_h
+#define INCLUDED_ml_maths_CBoostedTreeFactory_h
+
+#include <core/CContainerPrinter.h>
+#include <core/CDataFrame.h>
+#include <core/CPackedBitVector.h>
+
+#include <maths/CBayesianOptimisation.h>
+#include <maths/CBoostedTree.h>
+
+#include <numeric>
+#include <sstream>
+#include <utility>
+#include <vector>
+
+namespace ml {
+namespace maths {
+
+class CNode;
+class CBoostedTreeImpl;
+
+//! SimpleFactory for CBoostedTree object
+class CBoostedTreeFactory final {
+public:
+    using TCBoostedTreeUPtr = std::unique_ptr<CBoostedTree>;
+
+public:
+    //! Construct a boosted tree object from parameters
+    static CBoostedTreeFactory constructFromParameters(std::size_t numberThreads,
+                                                       std::size_t dependentVariable,
+                                                       CBoostedTree::TLossFunctionUPtr loss);
```

Comment:
nit: why not just a constructor?

Reply:
There can be potentially another method for construction which takes to `size_t`s and a loss function. Hence, I have a descriptive method name to identify the intention.

Reply:
Ok, I understand you intention. My argument against this is to me whatever is passed to this function should be required (not possible to default) and therefore there should only ever be one such function. If some of the values are optional they should be set by setter on the factory as we do for all the other optional parameters. I'm not 100% against this, and it is after all a matter of style, it just feels a little cumbersome to me.